odilia/src/tower/service_ext.rs

```diff
@@ -0,0 +1,44 @@
+use crate::tower::{
+	async_try::{AsyncTryInto, AsyncTryIntoLayer, AsyncTryIntoService},
+	iter_svc::IterService,
+	state_svc::{StateLayer, StateService},
```

Comment:
does state have to be a service? isn't it an extractor?

Reply:
> does state have to be a service?

yes

Named poorly, it could be more generic like "InjectorService"; the "StateService" really just clones an `Arc<S>` into its inner service... that's it.

> isn't it an extractor?

No

The "extractor" portion is done via the `TryFromState` trait, which takes an `S` and `T` and produces `Self`. This way of doing it enables any kind of state to be extracted, and some types of state not being accessible to certain kinds of handlers. You mentioned you wanted some kind of safety in this regard in a Matrix chat.

Take a look at the following snippet:

```rust
impl<C> TryFromState<Arc<ScreenReaderState>, C> for Speech
where
  C: CommandType + Debug,
{
  type Error = OdiliaError;
  type Future = Ready<Result<Speech, Self::Error>>;
  fn try_from_state(state: Arc<ScreenReaderState>, _cmd: C) -> Self::Future {
    ok(Speech(state.ssip.clone()))
  }
}
```

This extractor effectively blocks atspi event handlers from taking in the SSIPClient and directly sending SSIP commands across the socket. It is only available to command handlers. This is due to the trait bound: `C: CommandType + ...`

Reply:
alright, I forgot we have command handlers, and there's no single entity which processes all collected commands. In that case, yes, those abstractions are sound